### PR TITLE
Build Script

### LINKED ISSUE #499: Script for building mainnet and testnet
**Is your feature request related to a problem? Please describe.**
It can be frustrating to build for different environments (mainnet, testnet, etc) because you need specific node endpoints and CSS files, and if you get those wrong it will cause errors during runtime that can be a bit cryptic if you are unfamiliar with Full Service.

**Describe the solution you'd like**
Add script(s) to the tools folder that builds mainnet and testnet binaries, including downloading the necessary consensus-enclave.css file. The ingest-enclave.css file isn't needed in the build step and can be downloaded exclusively in the run-<net> script.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by pouneh**
Create a build script that downloads sigstructs and builds full service for the specified version of the mobilecoin network

Resolves #499

### CODE REVIEW

File: tools/build-fs.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+NET=$1
+fi
+
+if [ $NET == "local" ]; then
+echo "Local networks are not currently supported with this script"
+exit 1
+elif [ $NET == "test" ]; then
+echo "Building Full Service with testnet settings"
+NAMESPACE="test"
+SM=SW
```
Comment (eranrund) [inline tools/build-fs.sh:19]: Can be simplified by just setting SGX_MODE/IAS_MODE

File: tools/build-fs.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+CONSENSUS_ENCLAVE_CSS="$WORK_DIR/consensus-enclave.css"
+
+(cd ${WORK_DIR} && CONSENSUS_SIGSTRUCT_URI=$(curl -s https://enclave-distribution.${NAMESPACE}.mobilecoin.com/production.json | grep consensus-enclave.css | awk '{print $2}' | tr -d \" | tr -d ,)
+curl -O https://enclave-distribution.${NAMESPACE}.mobilecoin.com/${CONSENSUS_SIGSTRUCT_URI})
+
+echo "building full service..."
+SGX_MODE=$SM \
+IAS_MODE=$IM \
+CONSENSUS_ENCLAVE_CSS=$CONSENSUS_ENCLAVE_CSS \
```
Comment (eranrund) [inline tools/build-fs.sh:43]: if you change the environment variable assignments to be prefixed with `export` they will make their way to cargo build and you wont have to re-iterate it here

File: tools/build-fs.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+cargo build --release -p mc-full-service
```
Comment (eranrund) [inline tools/build-fs.sh:44]: I'd remove the `-p mc-full-service`, is there a reason not to build the other binaries in the full-service workspace?

File: tools/build-fs.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+IM=PROD
+else
+echo "Network specified is not valid. Try 'test' or 'main'. Default is test"
+exit 1
+fi
+
+WORK_DIR="$HOME/.mobilecoin/${NET}"
+WALLET_DB_DIR="${WORK_DIR}/wallet-db"
+LEDGER_DB_DIR="${WORK_DIR}/ledger-db"
+mkdir -p ${WORK_DIR}
```
Comment (eranrund) [inline tools/build-fs.sh:33]: This seems unrelated to building full-service. Should this be part of a "run full service" script instead?

File: tools/build-fs.sh (hunk 3 above)
Comment (briancorbin) [reply tools/build-fs.sh:39]: Agreed, can remove the project specific tag

File: tools/build-fs.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+fi
+
+if [ $NET == "local" ]; then
+echo "Local networks are not currently supported with this script"
+exit 1
+elif [ $NET == "test" ]; then
+echo "Building Full Service with testnet settings"
+NAMESPACE="test"
+export SGX_MODE=SW
+export IAS_MODE=DEV
```
Comment (eranrund) [inline tools/build-fs.sh:20]: I missed this at a previous review, but testnet uses HW/PROD like mainnet.
Local network uses SW/DEV

File: tools/build-fs.sh [hunk 6]
```diff
@@ -1,39 +1,41 @@
...
+    NAMESPACE="test"
+    export SGX_MODE=HW
+    export IAS_MODE=PROD
+    elif [ $NET == "main" ]; then
+    NAMESPACE="prod"
+    export SGX_MODE=HW
+    export IAS_MODE=PROD
+    fi 
+    
+    WORK_DIR="$HOME/.mobilecoin/${NET}"
```
Comment (eranrund) [inline tools/build-fs.sh:33]: What will happen here if NET is not set?

**Review by eranrund — CHANGES_REQUESTED** [review]

File: tools/build-fs.sh [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+fi
+
+if [ -z ${NET} ] || [ $NET -ne "test" ] || [ $NET -ne "main" ]; then
+echo "Using current environment's SGX_MODE, IAS_MODE, CONSENSUS_ENCLAVE_CSS "
+else 
+    if [ $NET == "test" ]; then
+    echo "Building Full Service with testnet settings"
+    NAMESPACE="test"
+    export SGX_MODE=HW
+    export IAS_MODE=PROD
```
Comment (eranrund) [inline tools/build-fs.sh:26]: Nit: Needs to be indented

File: tools/build-fs.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+#   version of the following variables are used
+#   - SGX_MODE
+#   - IAS_MODE
+#   - CONSENSUS_ENCLAVE_CSS
+#   - NAMESPACE (if no namespace is present, we use 'default')
+
+
+# Net can be main/test/local
+if [ $# -gt 0 ]; then
+NET=$1
```
Comment (eranrund) [inline tools/build-fs.sh:15]: This should be indented.

File: tools/build-fs.sh (hunk 6 above)
Comment (eranrund) [reply tools/build-fs.sh:29]: IT appears that we can get here with `NET` not being set to anything, which would just set `WORK_DIR` to `$HOME/.mobilecoin/`

File: tools/build-fs.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+        export IAS_MODE=PROD
+    fi 
+    
+    WORK_DIR="$HOME/.mobilecoin/${NET}"
+    mkdir -p ${WORK_DIR}
+    export CONSENSUS_ENCLAVE_CSS="$WORK_DIR/consensus-enclave.css"
+    (cd ${WORK_DIR} && CONSENSUS_SIGSTRUCT_URI=$(curl -s https://enclave-distribution.${NAMESPACE}.mobilecoin.com/production.json | grep consensus-enclave.css | awk '{print $2}' | tr -d \" | tr -d ,)
+        curl -O https://enclave-distribution.${NAMESPACE}.mobilecoin.com/${CONSENSUS_SIGSTRUCT_URI})
+else
+    echo "Using current environment's SGX_MODE, IAS_MODE, CONSENSUS_ENCLAVE_CSS "
```
Comment (eranrund) [inline tools/build-fs.sh:35]:
```suggestion
    echo "Using current environment's SGX_MODE, IAS_MODE, CONSENSUS_ENCLAVE_CSS"
```

File: tools/build-fs.sh [hunk 10]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+# Copyright (c) 2018-2022 The MobileCoin Foundation
+
+# To use this script, run build-fs test or build-fs main
+# If no network is specified, or a different network is specified, the env's 
+#   version of the following variables are used
+#   - SGX_MODE
+#   - IAS_MODE
+#   - CONSENSUS_ENCLAVE_CSS
+#   - NAMESPACE (if no namespace is present, we use 'default')
```
Comment (eranrund) [inline tools/build-fs.sh:10]: I don't see the string `default` mentioned anywhere in this script other than this comment.
On this topic, what do we get from allowing this script to run without a namespace?

**Review by eranrund — CHANGES_REQUESTED** [review]

File: tools/build-fs.sh (hunk 6 above)
Comment (eranrund) [reply tools/build-fs.sh:29]: Ah, my mistake. I misread the `if` statement.
Looks like it can be simplified to just `[ $NET == "test" ] || [ $NET == "main" ]` since that already implies `$NET` is not empty

File: tools/build-fs.sh [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    WORK_DIR="$HOME/.mobilecoin/${NET}"
+    mkdir -p ${WORK_DIR}
+    export CONSENSUS_ENCLAVE_CSS="$WORK_DIR/consensus-enclave.css"
+    (cd ${WORK_DIR} && CONSENSUS_SIGSTRUCT_URI=$(curl -s https://enclave-distribution.${NAMESPACE}.mobilecoin.com/production.json | grep consensus-enclave.css | awk '{print $2}' | tr -d \" | tr -d ,)
+        curl -O https://enclave-distribution.${NAMESPACE}.mobilecoin.com/${CONSENSUS_SIGSTRUCT_URI})
+else
+    echo "Using current environment's SGX_MODE, IAS_MODE, CONSENSUS_ENCLAVE_CSS"
+fi
+
+exit 0
```
Comment (eranrund) [inline tools/build-fs.sh:36]: Shouldn't be here

**Review by eranrund — CHANGES_REQUESTED** [review]

File: tools/build-fs.sh (hunk 6 above)
Comment (eranrund) [reply tools/build-fs.sh:27]: This should be mentioned in a comment then.
Also, it would be helpful for me if conversations are not marked as resolved until I have a chance to look at them, otherwise I can't tell what really was addressed and what still has a followup (such as documenting this behavior in a comment). Thanks!

Comment (eranrund) [reply tools/build-fs.sh:27]: Unrelated, I believe that if this is changed to `"$NET" == "test"` then there won't be need for the `-z` test...

File: tools/build-fs.sh [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+#   - SGX_MODE
+#   - IAS_MODE
+#   - CONSENSUS_ENCLAVE_CSS
+
+# Net can be main/test/local
+if [ $# -gt 0 ]; then
+    NET=$1
+fi
+
+if ([ "$NET" == "test" ] || [ "$NET" == "main" ]); then
```
Comment (eranrund) [inline tools/build-fs.sh:16]: Nit: Outer parenthesis are likely not required

File: tools/build-fs.sh (hunk 6 above)
Comment (eranrund) [reply tools/build-fs.sh:27]: Thank you!

**Review by eranrund — APPROVED** [review]
LGTM once the nit is addressed

**Review by briancorbin — APPROVED** [review]
